### PR TITLE
Max Selections for Quizzes 

### PR DESCRIPTION
**Pull request by LuckyHariharan**
@vithushan19 Ready for review https://www.loom.com/share/80b52b0db2c64e53a91532e0edf2327d

Main take aways:
-Refactored to use the View/Data type structures
-Changed Progress bar from static values to dynamically calculated for each quiz
-Implemented some Testing for MaxSelection to ensure code confidence

**Max Selection**
- Add a maxSelections prop which disallows the user from over-selecting
- If maxSelections == 1, the user can freely click any option and it will automatically change for them

### CODE REVIEW

File: components/resources/quizzes/shared/SkillSelections.tsx [hunk 1]
```diff
@@ -18,15 +20,22 @@ const SkillSelections: React.FC<SkillSelectionsProps> = ({
...
 }) => {
   const [selected, setSelected] = useState<string[]>([]);
+  useEffect(() => {
+    setSelected([]);
+  }, [progress]);
   const handleSelection = (selection: string) => {
     if (selected.includes(selection)) {
       setSelected(selected.filter((s) => s !== selection));
-    } else {
+    } else if (maxSelections === 1) {
```
Comment (LuckyHariharan) [inline components/resources/quizzes/shared/SkillSelections.tsx:32]: edge case if maxSelections === 1

File: components/resources/quizzes/shared/SkillSelections.tsx [hunk 2]
```diff
@@ -18,15 +20,22 @@ const SkillSelections: React.FC<SkillSelectionsProps> = ({
   progress,
   title,
   body,
+  maxSelections,
 }) => {
   const [selected, setSelected] = useState<string[]>([]);
+  useEffect(() => {
```
Comment (LuckyHariharan) [inline components/resources/quizzes/shared/SkillSelections.tsx:26]: useEffect to clear the selections. @jimip85 you might find this helpful for your list datastrcture that captures all the users selections.

**Comment by LuckyHariharan** [comment]
Had a couple bugs because I was a little too ambitious in this PR and tried to keep track of each pages preferences. Will talk to Jimi to pair on that problem instead.

File: components/resources/quizzes/shared/SkillSelections.tsx [hunk 3]
```diff
@@ -1,14 +1,25 @@
-import React, { useState } from "react";
+import React, { useEffect, useState } from "react";
 import { Button } from "../../../ui/Button";
 import Progress from "./Progress";
 import SkillifyNavbar from "./SkillifyNavbar";
+
+type QuizObject = {
+  industries: string[];
```
Comment (vithushan19) [inline components/resources/quizzes/shared/SkillSelections.tsx:7]: This shouldn't be in shared, this is specific to your quiz not all quizzes

File: components/resources/quizzes/shared/SkillSelections.tsx [hunk 4]
```diff
@@ -18,15 +29,38 @@ const SkillSelections: React.FC<SkillSelectionsProps> = ({
...
   body,
+  maxSelections,
+  setQuizObject,
+  quizObject,
+  page,
 }) => {
   const [selected, setSelected] = useState<string[]>([]);
+
+  // Save selected skills to quizObject whenever selected changes
+  useEffect(() => {
```
Comment (vithushan19) [inline components/resources/quizzes/shared/SkillSelections.tsx:40]: Nice

File: components/resources/quizzes/shared/SkillSelections.tsx [hunk 5]
```diff
@@ -18,15 +29,38 @@ const SkillSelections: React.FC<SkillSelectionsProps> = ({
...
+      return {
+        ...prev,
+        [page]: selected,
+      };
+    });
+  }, [selected]);
+
+  // Restore selected skills from quizObject whenever the component mounts
+  useEffect(() => {
+    setSelected(quizObject[page] || []);
```
Comment (vithushan19) [inline components/resources/quizzes/shared/SkillSelections.tsx:51]: Nice!

File: components/resources/quizzes/shared/SkillSelections.tsx [hunk 6]
```diff
@@ -18,15 +29,38 @@ const SkillSelections: React.FC<SkillSelectionsProps> = ({
...
+  }, [progress]);
+
   const handleSelection = (selection: string) => {
     if (selected.includes(selection)) {
       setSelected(selected.filter((s) => s !== selection));
-    } else {
+    } else if (maxSelections === 1) {
+      setSelected([selection]);
```
Comment (vithushan19) [inline components/resources/quizzes/shared/SkillSelections.tsx:58]: Amazing

File: pages/career-quiz.tsx [hunk 7]
```diff
@@ -27,6 +27,13 @@ const CareerQuiz = () => {
     setStage((prevStage) => prevStage - 1);
   };
 
+  // Define the object and initialize it with empty values
+  const [quizObject, setQuizObject] = useState({
+    industries: [],
```
Comment (vithushan19) [inline pages/career-quiz.tsx:32]: This is fine here, but I think we don't have to specify this type in the shared component with these exact keys.

In the shared component you can use the Record type in typescript.

Search for other uses of the Record type in our repo

File: components/resources/quizzes/shared/SkillSelections.tsx (hunk 3 above)
Comment (vithushan19) [reply components/resources/quizzes/shared/SkillSelections.tsx:7]: I think the type of quiz object should be a record of string to an array of some other type.

Try to figure out how record can help, if you get stuck we can chat this week

File: components/resources/quizzes/shared/SkillSelections.tsx (hunk 6 above)
Comment (vithushan19) [reply components/resources/quizzes/shared/SkillSelections.tsx:58]: Can we test this somehow? Not sure if this is easy to test with jest

Comment (vithushan19) [reply components/resources/quizzes/shared/SkillSelections.tsx:58]: Or a testing library

**Review by vithushan19 — COMMENTED** [review]
Looking good. Can we add some tests

**Review by vithushan19 — CHANGES_REQUESTED** [review]
this would be a good change to get in if you can address earlier comments. @LuckyHariharan

File: components/resources/quizzes/careerQuiz/career-quiz.tsx
Comment (jimip85) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:1]: Awesome!

File: components/resources/quizzes/careerQuiz/career-quiz.tsx [hunk 8]
```diff
@@ -50,7 +55,7 @@ const CareerQuiz = () => {
         };
       }),
       currentQuestion: 0,
-      progress: 0,
+      progress: quizData.questions.length,
```
Comment (vithushan19) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:58]: I'm pretty sure this useEffect is not needed, and you can just use this object in the state initializer at line 38

File: components/resources/quizzes/careerQuiz/career-quiz.tsx [hunk 9]
```diff
@@ -83,14 +93,50 @@ const CareerQuiz = () => {
...
+    const currentQuestion =
+      quizViewState.questions[quizViewState.currentQuestion];
+    const selectedOptions = currentQuestion.options.filter(
+      (option) => option.isSelected
+    );
+    const maxSelection =
+      quizData.questions[quizViewState.currentQuestion].maxSelections;
+    // MaxSelection ===1 edgecase
+    if (maxSelection === 1) {
+      // Set all other options to false and set the selected option to true
```
Comment (vithushan19) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:113]: great comment!

File: components/resources/quizzes/careerQuiz/career-quiz.tsx [hunk 10]
```diff
@@ -83,14 +93,50 @@ const CareerQuiz = () => {
...
+      const updatedQuizViewState = {
+        ...quizViewState,
+        questions: selectedQuizOption,
+      };
+      setQuizViewState(updatedQuizViewState);
+      return;
+    } else if (selectedOptions.length >= maxSelection && !option.isSelected) {
+      return;
+    }
     const selectedQuizOption = quizViewState.questions.map((question) => ({
```
Comment (vithushan19) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:132]: I might wrap the rest of this code in an else statement, if that makes sense to you.

File: components/resources/quizzes/shared/Progress.tsx [hunk 11]
```diff
@@ -1,14 +1,16 @@
...
 
-const Progress: React.FC<ProgressProps> = ({ progress }) => {
+const Progress: React.FC<ProgressProps> = ({ progress, index }) => {
   return (
     <div className="relative w-full h-4 bg-gray-300 rounded-full">
       <div
         className="absolute top-0 left-0 h-full rounded-full bg-violet-400"
-        style={{ width: `${progress}%` }}
+        // custom progress bar styling, so that the progress bar automatically starts at zero and evenly increments to 100 at the BluePrint
+        style={{ width: `${index >= 0 ? (index * 100) / progress : 100}%` }}
```
Comment (vithushan19) [inline components/resources/quizzes/shared/Progress.tsx:13]: I think all this calculation should be happening in the view state, not in the component. `progress` should be a percentage that you just pass in. This component shouldnt need to know the current index. You're just passing that in here to do you calculation. If you do the calculation in the state, it's easiest to test as well.

File: components/resources/quizzes/careerQuiz/career-quiz.tsx (hunk 8 above)
Comment (vithushan19) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:58]: progress should be a percent between 0 and 100 that you can just pass into the progress bar. Do the calculations in the state management, not in the progress bar view.

File: components/resources/quizzes/shared/SkillSelection.test.tsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  };
+
+  const handleOptionClickMock = jest.fn(); // create mock function
+
+  const { getByText } = render(
+    <SkillSelections
+      onNextClick={onNextClick}
+      onBackClick={onBackClick}
+      handleOptionClick={handleOptionClickMock} // pass mock function
+      quizViewState={quizViewState}
```
[the remark this answers is not in the document]
Comment (vithushan19) [reply components/resources/quizzes/shared/SkillSelection.test.tsx:24]: sounds good, feel free to send me a loom if you get stuck

**Review by vithushan19 — CHANGES_REQUESTED** [review]
looking really good! nice strong PR. clean and readable code, great comments. Keep it up!

File: components/resources/quizzes/shared/SkillSelection.test.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (LuckyHariharan) [reply components/resources/quizzes/shared/SkillSelection.test.tsx:24]: Resolved!

**Pull request closed without merging**